big-number/src/main/java/BigNumber.java

```diff
@@ -0,0 +1,230 @@
+            num2 = addSomeZeros(num2, num1.length());
+        }
+        if (num1.length() < num2.length()) {
+            num1 = addSomeZeros(num1, num2.length());
+        }
+
+        int addUnits = 0;
+        int subTens = 0;
+        for (int i = num1.length() - 1; i >= 0; i--) {
+            int next = charToDigit(num1.charAt(i)) - subTens - charToDigit(num2.charAt(i));
+            if (next < 0) {
+                subTens = 1;
+                addUnits = 10;
+            } else {
+                subTens = 0;
+            }
+            next += addUnits;
+            resultSB.append(next);
+            addUnits = 0;
+        }
+
+        int i = resultSB.length() - 1;
+        while (resultSB.charAt(i) == '0' && resultSB.length() > 1) {
+            resultSB.deleteCharAt(i);
+            i--;
+        }
+        resultSB.trimToSize();
+        return resultSB.reverse().toString();
+    }
+
+    public BigNumber plus(BigNumber num) {
+
+        // this and num are both positive
+        if (this.positive && num.positive) {
+            String thisPlusNum = digitsPlus(this.absValue, num.absValue);
+            return new BigNumber(thisPlusNum, true);
+        }
+
+        // this and num are both negative
+        if (!this.positive && !num.positive) {
+            String thisPlusNum = digitsPlus(this.absValue, num.absValue);
+            return new BigNumber(thisPlusNum, false);
+        }
+
+        BigNumber absThis = new BigNumber(this.absValue, true);
+        BigNumber absNum = new BigNumber(num.absValue, true);
+        int compareAbsValuesInt = absThis.compareTo(absNum);
+
+        // abs_this is bigger than abs_num
+        if (compareAbsValuesInt > 0) {
+            String thisMinusNum = digitsMinus(this.absValue, num.absValue);
+            return (this.positive) ? new BigNumber(thisMinusNum, true) : new BigNumber(thisMinusNum, false);
+        }
+        // abs_this smaller than abs_num
+        if (absThis.compareTo(absNum) < 0) {
+            String numMinusThis = digitsMinus(num.absValue, this.absValue);
+            return (this.positive) ? new BigNumber(numMinusThis, false) : new BigNumber(numMinusThis, true);
+        }
+        // |this| == |num|, but one of it is negative, the other - positive
+        return BIG_NUMBER_ZERO;
+    }
+
+
+    public BigNumber minus(BigNumber num) {
+
+        // this is positive and num is negative
+        if (this.positive && !num.positive) {
+            String thisPlusNum = digitsPlus(this.absValue, num.absValue);
+            return new BigNumber(thisPlusNum, true);
+        }
+
+        // this is negative and num is positive
+        if (!this.positive && num.positive) {
+            String thisPlusNum = digitsPlus(this.absValue, num.absValue);
+            return new BigNumber(thisPlusNum, false);
+        }
+
+        BigNumber absThis = new BigNumber(this.absValue, true);
+        BigNumber absNum = new BigNumber(num.absValue, true);
```

Comment:
I remember we discussed about a private method to compare abs values here.
It's up to you, because constructor doesn't do validation anymore.

Reply:
thats why I decided to leave it like this, because new constructor is very simple)